sdlib/d/gc/extent.d

```diff
@@ -306,57 +306,127 @@ public:
 	}
 
 	/**
-	 * Freespace Flag features for slabs.
+	 * Metadata features for slabs.
 	 */
 
 	@property
-	bool supportsFreeSpace() const {
+	bool supportsSlabMetaData() const {
 		return isAppendableSizeClass(sizeClass);
 	}
 
 	@property
-	ref shared(Bitmap!256) freeSpaceFlags() {
-		assert(isSlab(), "freeSpaceFlags accessed on non slab!");
-		assert(supportsFreeSpace, "freeSpaceFlags not supported!");
+	ref shared(Bitmap!256) slabMetaDataFlags() {
+		assert(isSlab(), "slabMetaDataFlags accessed on non slab!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
 
-		return _metadata.slabData.freeSpaceData.freeSpaceFlags;
+		return _metadata.slabData.slabMetaData.slabMetaDataFlags;
+	}
+
+	bool hasSlabMetaData(uint index) {
+		assert(isSlab(), "hasSlabMetaData accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		return supportsSlabMetaData && slabMetaDataFlags.valueAtAtomic(index);
 	}
 
 	ushort* freeSpacePtr(uint index) {
 		assert(isSlab(), "freeSpacePtr accessed on non slab!");
-		assert(supportsFreeSpace, "size class not supports freeSpace!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
 
 		return cast(ushort*) (address + (index + 1) * slotSize) - 2;
 	}
 
 	void setFreeSpace(uint index, size_t freeSpace) {
 		assert(isSlab(), "setFreeSpace accessed on non slab!");
-		assert(freeSpace <= slotSize, "freeSpace exceeds alloc size!");
+		assert(freeSpace <= getTotalCapacity(index),
+		       "freeSpace exceeds alloc size!");
 		assert(index < slotCount, "index is out of range!");
-		assert(supportsFreeSpace, "size class not supports freeSpace!");
+		assert(supportsSlabMetaData, "size class not supports slab metadata!");
 
-		if (freeSpace == 0) {
-			freeSpaceFlags.clearBitAtomic(index);
+		// If we have a finalizer, the freeSpace may be 0:
+		if (freeSpace == 0 && !hasFinalizer(index)) {
+			slabMetaDataFlags.clearBitAtomic(index);
 			return;
 		}
 
 		// Encode freespace and write it to the last byte (or two bytes) of alloc.
 		writePackedFreeSpace(freeSpacePtr(index), freeSpace & ushort.max);
-		freeSpaceFlags.setBitAtomic(index);
+		slabMetaDataFlags.setBitAtomic(index);
 	}
 
 	size_t getFreeSpace(uint index) {
 		assert(isSlab(), "getFreeSpace accessed on non slab!");
 		assert(index < slotCount, "index is out of range!");
 
-		if (!supportsFreeSpace || !freeSpaceFlags.valueAtAtomic(index)) {
+		if (!hasSlabMetaData(index)) {
 			return 0;
 		}
 
 		// Decode freespace, found in the final byte (or two bytes) of the alloc:
 		return readPackedFreeSpace(freeSpacePtr(index));
 	}
 
+	size_t getTotalCapacity(uint index) {
```

Comment:
I'm not sure why the capacity has to be total. What else could it be? If you insist on qualifying, then maybe getSlotCapacity, but getCapacity is probably fine, the index kinda gives it away that it's about a specific slot.